### PR TITLE
Storyteller update. Better pop thresholds. Pop scalling. Msg pref and tweaks

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
Increases the minimum thresholds in order for major and roleset (antag) events to happen.

Added a population scalling mechanic that will scale the event frequencies based upon the amount of people. It compares the population count between the threshold to run events and threshold to stop applying the penalty and applies up to 30% penalty. Both those values are configurable and this can be disabled too.

How event running with frequencies works with the scalling:
At 4 pop mundane events start to appear, but they have -30% penalty to the frequency
At 7 pop mundane events run with -15% penalty
At 10+ pop mundane events run with no penalty

At 20 pop major events start to appear, but they have -30% penalty to the frequency
At 30 pop major events run with -15% penalty
At 40+ pop major events run with no penalty

Admins can now disable getting messages about storyteller operations through a pref.

Medical moderate events like appendicitis and brain trauma has gotten their weight lowered
False alarm has gotten its weight lowered too.

Various small code tweaks and improvements.

## Changelog
:cl:
add: The storyteller system has been updated check https://github.com/hrzntal/horizon/pull/617 for more info.
/:cl:

### CODE REVIEW

File: config/game_options.txt [hunk 1]
```diff
@@ -529,13 +529,30 @@ OBJECTIVES_ROUNDSTART_POINT_MULTIPLIER 1
...
 
 ## Point thresholds for tracks to run events. The lesser the more frequent events will be.
 MUNDANE_POINT_THRESHOLD 25
 MODERATE_POINT_THRESHOLD 50
 MAJOR_POINT_THRESHOLD 90
 ROLESET_POINT_THRESHOLD 120
 OBJECTIVES_POINT_THRESHOLD 130
+
+## Allows the storyteller to scale event frequencies based on population
+ALLOW_STORYTELLER_POP_SCALLING
```
Comment (atakiya) [inline config/game_options.txt:544]:
```suggestion
ALLOW_STORYTELLER_POP_SCALING
```

File: code/controllers/subsystem/gamemode.dm [hunk 2]
```diff
@@ -950,6 +1038,10 @@ SUBSYSTEM_DEF(gamemode)
 				if("vars")
 					var/track = href_list["track"]
 					switch(href_list["var"])
+						if("allow_pop_scalling")
+							allow_pop_scalling = !allow_pop_scalling
+							if(allow_pop_scalling)
+								update_pop_scalling()
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:1044]:
```suggestion
						if("allow_pop_scaling")
							allow_pop_scaling = !allow_pop_scaling
							if(allow_pop_scaling)
								update_pop_scaling()
```

File: code/controllers/subsystem/gamemode.dm [hunk 3]
```diff
@@ -761,6 +826,26 @@ SUBSYSTEM_DEF(gamemode)
...
+				else
+					multiplier = 0
+				dat += "[track]: -[multiplier]% "
+			dat += "<BR><b>Population Scale Thresholds:</b>"
+			dat += "<BR><font color='#888888'><i>Those are thresholds at which the population scalling penalties for event frequency no longer will apply.</i></font>"
+			for(var/track in event_tracks)
+				dat += "<BR>[track]: <a href='?src=[REF(src)];panel=main;action=vars;var=pop_scale_threshold;track=[track]'>[pop_scale_thresholds[track]]</a>"
+
+			dat += "<BR><b>Population Scale Penalties:</b>"
+			dat += "<BR><font color='#888888'><i>Those are maximum penalties for event frequencies from population scalling. Calculates as percentages.</i></font>"
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:845]:
```suggestion
			dat += "<BR><font color='#888888'><i>Those are maximum penalties for event frequencies from population scaling. Calculates as percentages.</i></font>"
```

File: code/controllers/subsystem/gamemode.dm [hunk 4]
```diff
@@ -761,6 +826,26 @@ SUBSYSTEM_DEF(gamemode)
...
+			dat += "<BR>Current pop penalties: "
+			for(var/track in event_tracks)
+				var/multiplier
+				if(allow_pop_scalling)
+					multiplier =  (1 - current_pop_scale_multipliers[track]) * 100
+				else
+					multiplier = 0
+				dat += "[track]: -[multiplier]% "
+			dat += "<BR><b>Population Scale Thresholds:</b>"
+			dat += "<BR><font color='#888888'><i>Those are thresholds at which the population scalling penalties for event frequency no longer will apply.</i></font>"
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:840]:
```suggestion
			dat += "<BR><font color='#888888'><i>Those are thresholds at which the population scaling penalties for event frequency no longer will apply.</i></font>"
```

File: code/controllers/subsystem/gamemode.dm [hunk 5]
```diff
@@ -761,6 +826,26 @@ SUBSYSTEM_DEF(gamemode)
 			for(var/track in event_tracks)
 				dat += "<BR>[track]: <a href='?src=[REF(src)];panel=main;action=vars;var=pts_threshold;track=[track]'>[point_thresholds[track]]</a>"
 
+			dat += "<HR><b>Allow Storyteller Pop Scalling:</b> <a href='?src=[REF(src)];panel=main;action=vars;var=allow_pop_scalling'>[allow_pop_scalling ? "Yes" : "No"]</a>"
+			dat += "<BR><font color='#888888'><i>This makes the events happens less frequency the less population there is, up to a cap.</i></font>"
+			dat += "<BR>Current pop penalties: "
+			for(var/track in event_tracks)
+				var/multiplier
+				if(allow_pop_scalling)
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:834]:
```suggestion
				if(allow_pop_scaling)
```

File: code/controllers/subsystem/gamemode.dm [hunk 6]
```diff
@@ -761,6 +826,26 @@ SUBSYSTEM_DEF(gamemode)
 			for(var/track in event_tracks)
 				dat += "<BR>[track]: <a href='?src=[REF(src)];panel=main;action=vars;var=pts_threshold;track=[track]'>[point_thresholds[track]]</a>"
 
+			dat += "<HR><b>Allow Storyteller Pop Scalling:</b> <a href='?src=[REF(src)];panel=main;action=vars;var=allow_pop_scalling'>[allow_pop_scalling ? "Yes" : "No"]</a>"
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:829]:
```suggestion
			dat += "<HR><b>Allow Storyteller Pop Scaling:</b> <a href='?src=[REF(src)];panel=main;action=vars;var=allow_pop_scaling'>[allow_pop_scaling ? "Yes" : "No"]</a>"
```

File: code/controllers/subsystem/gamemode.dm [hunk 7]
```diff
@@ -660,6 +710,9 @@ SUBSYSTEM_DEF(gamemode)
 
 /// Loads config values from game_options.txt
 /datum/controller/subsystem/gamemode/proc/load_config_vars()
+	halted_storyteller = CONFIG_GET(flag/halt_storyteller)
+	allow_pop_scalling = CONFIG_GET(flag/allow_storyteller_pop_scalling)
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:714]:
```suggestion
	allow_pop_scaling = CONFIG_GET(flag/allow_storyteller_pop_scaling)
```

File: code/controllers/subsystem/gamemode.dm [hunk 8]
```diff
@@ -347,6 +375,28 @@ SUBSYSTEM_DEF(gamemode)
 				med_crew++
 			if(player_role.departments_bitflags & DEPARTMENT_BITFLAG_SECURITY)
 				sec_crew++
+	if(allow_pop_scalling)
+		update_pop_scalling()
+
+/datum/controller/subsystem/gamemode/proc/update_pop_scalling()
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:381]:
```suggestion
	if(allow_pop_scaling)
		update_pop_scaling()

/datum/controller/subsystem/gamemode/proc/update_pop_scaling()
```

File: code/controllers/subsystem/gamemode.dm [hunk 9]
```diff
@@ -66,6 +66,36 @@ SUBSYSTEM_DEF(gamemode)
 		EVENT_TRACK_OBJECTIVES = 1
 		)
 
+	/// Whether we allow pop scalling. This is configured by config, or the storyteller UI
+	var/allow_pop_scalling = TRUE
```
Comment (atakiya) [inline code/controllers/subsystem/gamemode.dm:70]:
```suggestion
	var/allow_pop_scaling = TRUE
```

File: code/controllers/configuration/entries/game_options.dm [hunk 10]
```diff
@@ -301,6 +301,8 @@
 
 /datum/config_entry/flag/halt_storyteller // Stops storyteller from playing random events and spawning antags. This can be freely toggled in game in the storyteller menu
 
+/datum/config_entry/flag/allow_storyteller_pop_scalling // Allows storyteller to scale down the event frequency by population
```
Comment (atakiya) [inline code/controllers/configuration/entries/game_options.dm:304]:
```suggestion
/datum/config_entry/flag/allow_storyteller_pop_scaling // Allows storyteller to scale down the event frequency by population
```

**Review by atakiya — CHANGES_REQUESTED** [review]
hope I caught all of them

**Comment by Azarak** [comment]
oof my grammar

**Review by atakiya — APPROVED** [review]

**Pull request merged**